### PR TITLE
Add `ignoreLonghands: []` to `declaration-block-no-redundant-longhand-properties`

### LINKED ISSUE #7232: Add `"ignoreLonghands": []` to `declaration-block-no-redundant-longhand-properties`
## use cases

- [`text-decoration`](#text-decoration)
- [`background`](#background)

## rationale

If a browser doesn't support the new arity for a given shorthand, it will drop it altogether.
For example if we were to introduce an enhancement for the current fixer of `text-decoration` by adding support to
`text-decoration-thickness`, for a lot of users of the rule it would introduce a hard to detect regression.  
They would have to rely on `"ignoreShorthands": ["text-decoration"]` which is arguably a downgrade compared to the status quo. Hence a new option is required to be more granular and cross-browser.

### rationale for new fix

https://github.com/stylelint/stylelint/pull/6580#discussion_r1095994177
#7060

### without option

```css
a {
  text-decoration-line: underline;
  text-decoration-style: solid;
  text-decoration-color: green;
  text-decoration-thickness: 1px;
}
```
becomes
```css
a {
  text-decoration: underline solid green 1px;
}
```

### with `"ignoreLonghands": ["text-decoration-thickness"]`

```css
a {
  text-decoration-line: underline;
  text-decoration-style: solid;
  text-decoration-color: green;
  text-decoration-thickness: 1px;
}
```
becomes
```css
a {
	text-decoration: underline solid green;
	text-decoration-thickness: 1px;
}
```

### references

https://bugs.webkit.org/show_bug.cgi?id=190774
https://drafts.csswg.org/css-text-decor-4/#text-decoration-property

### implementation details

[Firefox 69 supports this feature (preffed off) under the name `text-decoration-width`, behind the pref layout.css.text-decoration-width.enabled](https://bugzilla.mozilla.org/show_bug.cgi?id=1567282#c8).
[`text-decoration-skip` is not part of the shorthand](https://drafts.csswg.org/css-text-decor-4/#text-decoration-property) but will have to be eventually added in a separate PR. 

### browser support

https://caniuse.com/?search=text-decoration-thickness

### rationale

https://caniuse.com/background-img-opts

### example

```json
{
  "fix": true,
  "rules": {
    "declaration-block-no-redundant-longhand-properties": [
      true,
      { "ignoreLonghands": ["background-size", "background-origin", "background-clip"] }
    ],
  }
}
```

```css
a {
  background-image: none;
  background-position: 0% 0%;
  background-size: auto auto;
  background-repeat: repeat;
  background-origin: padding-box;
  background-clip: border-box;
  background-attachment: scroll;
  background-color: transparent;
}
```
becomes
```css
a {
  background: repeat scroll 0% 0% none transparent;
  background-size: auto auto;
  background-origin: padding-box;
  background-clip: border-box;
}
```

### PR DESCRIPTION
**Pull request by Mouvedia**
> Which issue, if any, is this issue related to?

Closes #7232

> Is there anything in the PR that needs further explanation?

To understand why a fix is being accompanied by an option check the discussion starting with https://github.com/stylelint/stylelint/issues/7232#issuecomment-1762654301

tl;dr: CSS Text Decoration Module Level 3 vs CSS Text Decoration Module Level 4 means that a granular option is necessary for the `text-decoration` fix not be considered a downgrade

### CODE REVIEW

File: lib/rules/declaration-block-no-redundant-longhand-properties/__tests__/index.mjs [hunk 1]
```diff
@@ -387,6 +395,50 @@ testRule({
...
+		{
+			code: 'a { text-decoration-line: underline; text-decoration-style: solid; text-decoration-color: purple; }',
+			fixed: 'a { text-decoration: underline solid purple; }',
+			description: 'CSS Text Decoration Module Level 3',
+			message: messages.expected('text-decoration'),
+		},
+		{
+			code: 'a { text-decoration-color: purple; text-decoration-thickness: 1px; text-decoration-style: solid; text-decoration-line: underline; text-decoration-width: 1px; }',
+			fixed:
+				'a { text-decoration: underline solid purple; text-decoration-thickness: 1px; text-decoration-width: 1px; }',
```
Comment (Mouvedia) [inline lib/rules/declaration-block-no-redundant-longhand-properties/__tests__/index.mjs:432]: I chose not to add an exception for `text-decoration-width` because #7510 doesn't have a resolution.

File: .changeset/sharp-buses-drop.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+---
+"stylelint": minor
+---
+
+Added: `ignoreLonghands` option to `declaration-block-no-redundant-longhand-properties`
```
Comment (ybiquitous) [inline .changeset/sharp-buses-drop.md:5]: [suggest] Let's keep consistency in CHANGELOG:

```suggestion
Added: `ignoreLonghands: []` to `declaration-block-no-redundant-longhand-properties`
```

File: lib/rules/declaration-block-no-redundant-longhand-properties/README.md [hunk 3]
```diff
@@ -174,6 +174,60 @@ a {
 
 ## Optional secondary options
 
+### `ignoreLonghands: ["string"]`
```
Comment (ybiquitous) [inline lib/rules/declaration-block-no-redundant-longhand-properties/README.md:177]: [suggest] I feel supporting also regexes makes sense. Any thoughts?

```suggestion
### `ignoreLonghands: ["/regex/", /regex/, "string"]`
```

File: lib/rules/declaration-block-no-redundant-longhand-properties/__tests__/index.mjs [hunk 4]
```diff
@@ -387,6 +395,50 @@ testRule({
...
+			ignoreLonghands: [
+				'background-size',
+				'background-origin',
+				'background-clip',
+				'text-decoration-thickness',
+			],
+		},
+	],
+	fix: true,
+
```
Comment (ybiquitous) [inline lib/rules/declaration-block-no-redundant-longhand-properties/__tests__/index.mjs:421]: [suggest] Can you add `accept` cases, too?

**Review by ybiquitous — CHANGES_REQUESTED** [review]
Thanks for the pull request. Can you address my comments?

File: .changeset/selfish-plums-dream.md [hunk 5]
```diff
@@ -0,0 +1,5 @@
+---
+"stylelint": minor
```
[the remark this answers is not in the document]
Comment (Mouvedia) [reply .changeset/selfish-plums-dream.md:2]: > Typo?

Nop

> [a bug fix in a rule that results in Stylelint reporting more errors](https://stylelint.io/about/semantic-versioning/#minor-release)

i.e. without the option

```css
a {
  text-decoration-line: underline;
  text-decoration-style: solid;
  text-decoration-color: purple;
  text-decoration-thickness: 1px; /* this one will newly be reported */
}

File: lib/rules/declaration-block-no-redundant-longhand-properties/README.md (hunk 3 above)
Comment (Mouvedia) [reply lib/rules/declaration-block-no-redundant-longhand-properties/README.md:177]: I don't think so for 2 reasons:

- it will complicate the implementation
- I couldn't find a use case for regexes

File: lib/rules/declaration-block-no-redundant-longhand-properties/__tests__/index.mjs (hunk 4 above)
Comment (Mouvedia) [reply lib/rules/declaration-block-no-redundant-longhand-properties/__tests__/index.mjs:421]: I couldn't think of good ones.
i.e. the accept ones needs to help with potential regressions

Do you have a suggestion?

Comment (ybiquitous) [reply lib/rules/declaration-block-no-redundant-longhand-properties/__tests__/index.mjs:421]: I see. Let's keep it.

File: lib/rules/declaration-block-no-redundant-longhand-properties/README.md (hunk 3 above)
Comment (ybiquitous) [reply lib/rules/declaration-block-no-redundant-longhand-properties/README.md:177]: I see. Let's leave it as-is. We can revisit later if people request.

File: lib/rules/declaration-block-no-redundant-longhand-properties/index.mjs [hunk 6]
```diff
@@ -234,6 +235,8 @@ const rule = (primary, secondaryOptions, context) => {
 			}
 
 			for (const longhand of longhandProps) {
+				if (secondaryOptions?.ignoreLonghands?.includes(longhand)) continue;
```
Comment (ybiquitous) [inline lib/rules/declaration-block-no-redundant-longhand-properties/index.mjs:238]: [suggest] How about defining a local variable outside the loop like this?

```js
		if (!validOptions) {
			return;
		}

		/** @type {string[]} */
		const ignoreLonghands = secondaryOptions?.ignoreLonghands ?? [];
```

File: lib/rules/declaration-block-no-redundant-longhand-properties/index.mjs [hunk 7]
```diff
@@ -275,11 +278,17 @@ const rule = (primary, secondaryOptions, context) => {
...
 
-					const shorthandProps = longhandSubPropertiesOfShorthandProperties.get(shorthandProperty);
-					const prefixedShorthandData = Array.from(shorthandProps || [], (item) => prefix + item);
+					const shorthandProps = new Set(
+						longhandSubPropertiesOfShorthandProperties.get(shorthandProperty),
+					);
 
+					secondaryOptions?.ignoreLonghands?.forEach((/** @type {string} */ value) => {
+						if (shorthandProps.has(value)) shorthandProps.delete(value);
+					});
```
Comment (ybiquitous) [inline lib/rules/declaration-block-no-redundant-longhand-properties/index.mjs:287]: [suggest] Extracting a local variable `ignoreLonghands`

```js
/** @type {string[]} */
const ignoreLonghands = secondaryOptions?.ignoreLonghands ?? [];
```

can simplify this code like this:

```suggestion
					ignoreLonghands.forEach((value) => shorthandProps.delete(value));
```

Just `.delete` seems enough (`.has` seems unneeded). Any thoughts?

File: lib/rules/declaration-block-no-redundant-longhand-properties/index.mjs (hunk 6 above)
Comment (Mouvedia) [reply lib/rules/declaration-block-no-redundant-longhand-properties/index.mjs:238]: I wish we didn't have to cast it.
Can you take a look at #7608?

Comment (ybiquitous) [reply lib/rules/declaration-block-no-redundant-longhand-properties/index.mjs:238]: The idea of `meta.schema` sounds interesting, but casting is necessary for now, right?

**Review by ybiquitous — APPROVED** [review]
Thanks. LGTM 👍🏼

**Comment by Mouvedia** [comment]
@ybiquitous @jeddy3 I would advise to add `"ignoreLonghands": ["text-decoration-thickness"]` to `stylelint-config-standard`.

**Comment by ybiquitous** [comment]
> I would advise to add `"ignoreLonghands": ["text-decoration-thickness"]`

Hum, [this example on MDN](https://developer.mozilla.org/en-US/docs/Web/CSS/text-decoration#examples) makes sense, right?

```css
.thick {
  text-decoration: solid underline purple 4px; /* includes `text-decoration-thickness: 4px` */
}
```

**Comment by Mouvedia** [comment]
It does if you consider that 95%+ support on caniuse is enough. If not it would be better to add the option so that standard doesn't change.

**Comment by ybiquitous** [comment]
> It does if you consider that 95%+ support on caniuse is enough.

I think the percentage is enough for now. Let's keep things simple. If people complain, then we can revisit it.

**Comment by Mouvedia** [comment]
@ybiquitous 

```
error - 2024-04-18 12:59:34,223 -- Commit creating failed: {"detail":"Tokenless has reached GitHub rate limit. Please upload using a token: https://docs.codecov.com/docs/adding-the-codecov-token. Expected available in 136 seconds."}
Error: Codecov: Failed to properly create commit: The process '/home/runner/work/_actions/codecov/codecov-action/v4/dist/codecov' failed with exit code 1
```

**Comment by ybiquitous** [comment]
Have you merge the commit fca3269efcabe1291746a6832cd197e688712bdd into this branch?
